### PR TITLE
ML-MIPT

### PR DESCRIPTION
**Pull request by WandaHou**
This pull request adds code for studying cluster states and measurement-induced entanglement in quantum systems using Cirq. The implementation targets a 6×6 qubit grid with ancilla qubits for readout error mitigation and includes support for circuit construction and data collection.

Key components include:
• Cluster state circuit generation with Hadamard, CZ, and single-qubit rotations
• Basis rotations for measurement in X, Y, Z bases with configurable probe qubits
• Distance-dependent entanglement measurements between probe qubits
• Data collection framework with randomized compiling for error mitigation

The code uses Google’s Cirq framework and is designed to run on quantum processors through Cirq-Google. It builds circuits on a fixed 6×6 qubit grid, with a configurable probe separation distance (d = 3–6) and support for various measurement bases, enabling the study of entanglement properties in cluster states.

The testing framework ensures parameter validation and circuit structure without requiring actual quantum execution, making it suitable for CI/CD pipelines while preserving the ability to run on real quantum hardware.

### CODE REVIEW

**Comment by WandaHou** [comment]
This pull request adds code for studying measurement-induced entanglement in quantum systems using Cirq. The implementation targets a 6×6 qubit grid with ancilla qubits for readout error mitigation and includes full support for circuit construction, data collection, and analysis.

Key components include:
• Circuit generation with Hadamard, CZ, and single-qubit rotations
• Measurement in randomized X, Y, Z bases for shadow tomography
• Post-processing with error mitigation and state reconstruction
• Tensor network methods for analyzing entanglement properties such as negativity and von Neumann entropy

This code is used in a collaboration between UCSD (Yi-Zhuang You’s group), UC Berkeley (Ehud Altman’s group), and Google Quantum AI (Pedram Roushan’s group), as part of the effort to apply machine learning to measurement-induced phase transitions (ML for MIPT). The experiment is designed to run on Google’s Willow Pink superconducting quantum chip.

**Comment by WandaHou** [comment]
Hi, this PR implements the measurement-induced entanglement experiment targeting the Willow Pink quantum chip. It is part of our collaboration with the Berkeley (Ehud Altman) and Google Quantum AI (Pedram Roushan) teams, under the ML for MIPT project led by Yi-Zhuang You's group at UCSD.

All tests have passed. Please let me know if any changes are needed. Thank you for your time and review!

File: recirq/Cluster_state_MIPT/data_collect.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                            for r in result])
+                        torch.save(
+                            measurements,
+                            f'data/{folder_name}_d={d}/theta{theta_idx}/loop{i}/theta={theta_idx}_({b1},{b2}).pt')
+
+
+def main() -> None:
+    """Main function to run the data collection."""
+    # Replace these with your specific engine and processor details
+    eng = cg.get_engine('your-engine-id')
```
Comment (eliottrosenberg) [inline recirq/Cluster_state_MIPT/data_collect.py:80]: This is actually your GCP project ID.

**Review by eliottrosenberg — APPROVED** [review]
@dstrain115 do you want to take a look at this?

@WandaHou are you interested in making an .ipynb notebook in https://github.com/quantumlib/ReCirq/tree/master/docs to explain your experiment?

File: requirements.txt [hunk 2]
```diff
@@ -5,4 +5,5 @@ cirq-google>=1.2.0
 seaborn
 sphinx
 ipython
-black
\ No newline at end of file
+black
+torch>=1.9.0
```
Comment (dstrain115) [inline requirements.txt:9]: Please add this to recirq/cluster_state_mipt/extra-requirements.txt instead so it is not installed for all modules.
(If we still need this, see above comments)

File: recirq/Cluster_state_MIPT/cluster_state.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
+import cirq # type: ignore
+import cirq_google as cg # type: ignore
+import numpy as np # type: ignore
```
Comment (dstrain115) [inline recirq/Cluster_state_MIPT/cluster_state.py:3]: Why is this annotated # type: ignore?  You should be able to configure mypy to not follow external packages if that's the issue.

File: recirq/Cluster_state_MIPT/cluster_state.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+def get_two_qubits_6x6(d=6):
```
Comment (dstrain115) [inline recirq/Cluster_state_MIPT/cluster_state.py:5]: Please add a docstring for this function.
Also explain what d is (and maybe change it to a more descriptive name).
If d is the size of the grid, then also change the name of the function, since it is only 6x6 if d=6.

File: recirq/Cluster_state_MIPT/cluster_state.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        [cirq.GridQubit(8,9), cirq.GridQubit(8,10)]],
+        # extra ancilla pairs
+        # [[cirq.GridQubit(2,8), cirq.GridQubit(1,8)]]
+    ]
+    # get all qubits
+    all_qubits = np.unique(np.concatenate(
+        [np.array(qubits_matrix).flatten()]+[np.array(anc_pair_set).flatten() for anc_pair_set in anc_pairs]))
+    return qubits_matrix, probe_qubits, anc_pairs, all_qubits
+
+def get_circuit(qubits_matrix, theta, phi, probe_qubits, basis=[0,0], anc_pairs=[]):
```
Comment (dstrain115) [inline recirq/Cluster_state_MIPT/cluster_state.py:64]: Please add docstring.

File: recirq/Cluster_state_MIPT/cluster_state_test.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    circuit = cluster_state.get_circuit(
+        qubits_matrix=qubits_matrix,
+        theta=theta,
+        phi=phi,
+        probe_qubits=probe_qubits,
+        basis=basis,
+        anc_pairs=anc_pairs
+    )
+
+    assert isinstance(circuit, cirq.Circuit)
```
Comment (dstrain115) [inline recirq/Cluster_state_MIPT/cluster_state_test.py:66]: These are pretty weak assertions.  They just verify that any non-empty circuit is returned.
Can this be more rigorous?

File: recirq/Cluster_state_MIPT/data_collect.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+import cirq
+import cirq_google as cg
+import torch
+import numpy as np
+from recirq.Cluster_state_MIPT import get_circuit, get_two_qubits_6x6
+
+
+def collect_data(
+    eng: cg.Engine,
```
Comment (dstrain115) [inline recirq/Cluster_state_MIPT/data_collect.py:14]: nit: recommend engine instead of eng. 
Avoid abbreviations, as per:
https://google.github.io/styleguide/pyguide.html#316-naming

File: recirq/Cluster_state_MIPT/data_collect.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    if num_rc_circuits <= 0:
+        raise ValueError('num_rc_circuits must be greater than 0')
+
+    sampler_real = eng.get_processor(processor_id).get_sampler(
+        device_config_name=device_config_name)
+    sampler = sampler_real
+    theta_range = np.linspace(0, np.pi/2, 11)
+    phi = np.pi*(5/4)
+
+    for d in [5]:
```
Comment (dstrain115) [inline recirq/Cluster_state_MIPT/data_collect.py:45]: Why is this a for loop if it only has one value?

File: recirq/Cluster_state_MIPT/data_collect.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        qubits_matrix, probe_qubits, anc_pairs, all_qubits = get_two_qubits_6x6(d=d)
+        for theta_idx in range(len(theta_range)):
+            for i in range(5):  # iterations
```
Comment (dstrain115) [inline recirq/Cluster_state_MIPT/data_collect.py:48]: nit:  rather than "for i in range(5):  # iterations" use "for iterations in range(5):" to be self-documenting.

File: recirq/Cluster_state_MIPT/data_collect.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+                            cirq.merge_single_qubit_moments_to_phxz(circuit)
+                            for circuit in circuit_batch]
+                        result = sampler.run_batch(
+                            circuit_batch,
+                            repetitions=int(repetitions/num_rc_circuits))
+                        print(f'distance={d}, loop={i}, theta={theta_idx}, basis={b1},{b2}')
+                        measurements = torch.cat([
+                            torch.tensor(r[0].measurements['m']).to(int)
+                            for r in result])
+                        torch.save(
```
Comment (dstrain115) [inline recirq/Cluster_state_MIPT/data_collect.py:72]: Are we importing all of pytorch just to save a file?  Do we really need pytorch to save a matrix to a file?
Could we, instead, use cirq's json serialization instead, for example?

File: recirq/Cluster_state_MIPT/cluster_state.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
+import cirq # type: ignore
```
Comment (dstrain115) [inline recirq/Cluster_state_MIPT/cluster_state.py:1]: All of these files need to have an Apache copyright header.  See other files in ReCirq for an example.

File: recirq/Cluster_state_MIPT/__init__.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+"""Cluster state measurement implementation at MIPT.
```
[the remark this answers is not in the document]
Comment (WandaHou) [reply recirq/Cluster_state_MIPT/__init__.py]: Done.

File: requirements.txt (hunk 2 above)
Comment (WandaHou) [reply requirements.txt:9]: I have removed the torch installation, and using numpy to save collected data.

File: recirq/Cluster_state_MIPT/cluster_state.py (hunk 3 above)
Comment (WandaHou) [reply recirq/Cluster_state_MIPT/cluster_state.py]: Got it, # type: ignore is removed.

File: recirq/Cluster_state_MIPT/cluster_state.py (hunk 4 above)
Comment (WandaHou) [reply recirq/Cluster_state_MIPT/cluster_state.py]: “d” here refers to the spatial separation of the probe qubit pair. In our experiment, it can take values of 3, 4, 5, or 6 in a 6x6 grid. I’ve added a docstring to clarify this.

File: recirq/Cluster_state_MIPT/cluster_state.py (hunk 5 above)
Comment (WandaHou) [reply recirq/Cluster_state_MIPT/cluster_state.py]: I have added docstring.

File: recirq/Cluster_state_MIPT/cluster_state_test.py (hunk 6 above)
Comment (WandaHou) [reply recirq/Cluster_state_MIPT/cluster_state_test.py]: I’ve modified this test to build the cluster state circuit and verify that it is a constant-depth (14) circuit with the correct measurement bases on the probe qubits. Let me know if this approach is better.

File: recirq/Cluster_state_MIPT/data_collect.py (hunk 7 above)
Comment (WandaHou) [reply recirq/Cluster_state_MIPT/data_collect.py]: I have changed it to engine and also in other places.

File: recirq/Cluster_state_MIPT/data_collect.py (hunk 8 above)
Comment (WandaHou) [reply recirq/Cluster_state_MIPT/data_collect.py]: I changed this to "for d in [3, 4, 5, 6]:" to iterate over different probe qubit distances.

File: recirq/Cluster_state_MIPT/data_collect.py (hunk 9 above)
Comment (WandaHou) [reply recirq/Cluster_state_MIPT/data_collect.py]: Done.

File: recirq/Cluster_state_MIPT/data_collect.py (hunk 10 above)
Comment (WandaHou) [reply recirq/Cluster_state_MIPT/data_collect.py]: I’ve changed the code to use NumPy for saving the data. Previously, we used PyTorch because the next step in the research involved applying machine learning methods to detect quantum entanglement. However, since we’re now only presenting the circuit construction, I switched to NumPy to eliminate the need for installing PyTorch.

File: recirq/Cluster_state_MIPT/cluster_state.py (hunk 11 above)
Comment (WandaHou) [reply recirq/Cluster_state_MIPT/cluster_state.py]: I have added Apache copyright header with year 2025.

**Comment by WandaHou** [comment]
@dstrain115 All the issues should be addressed now—let me know if there’s anything else I should update!

File: requirements.txt (hunk 2 above)
Comment (dstrain115) [reply requirements.txt:9]: Let's remove this from the requirements then.

**Review by dstrain115 — COMMENTED** [review]
LGTM after the last comments.

File: recirq/cluster_state_mipt/__init__.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+"""Cluster state measurement implementation at MIPT.
```
Comment (dstrain115) [inline recirq/cluster_state_mipt/__init__.py:1]: I think even init.py are supposed to have copyright notice.

**Review by dstrain115 — APPROVED** [review]
LGTM, thanks for doing the changes!

**Comment by WandaHou** [comment]
It is telling me:

The failing test, test_size_independence_muinf, in the recirq/kpz/experiment_test.py file is caused by the np.isclose assertion for kurtosis.

What do you suggest me to do next step? @dstrain115

**Comment by dstrain115** [comment]
Could be a flaky test.  I re-ran it.

**Comment by eliottrosenberg** [comment]
@WandaHou I think that's my fault for writing a non-deterministic test. I can fix it by passing a fixed seed to the random number generator.

**Comment by eliottrosenberg** [comment]
@WandaHou I know I'm a little late in pointing this out, but you probably also want to update this file to add your work to the table of contents on the website: https://github.com/quantumlib/ReCirq/blob/master/docs/_index_included.yaml.

**Comment by eliottrosenberg** [comment]
Actually, I guess you never made a notebook for the `docs` folder, so never mind.
